orca/src/llm/openai.rs

```diff
@@ -273,25 +274,40 @@ impl EmbeddingTrait for OpenAI {
         Ok(res.into())
     }
 
-    /// TODO: Concurrent
-    async fn generate_embeddings(&self, prompt: Vec<Box<dyn Prompt>>) -> Result<EmbeddingResponse> {
+    async fn generate_embeddings(&self, prompts: Vec<Box<dyn Prompt>>) -> Result<EmbeddingResponse> {
         let mut embeddings = Vec::new();
-        for prompt in prompt {
+        let (sender, receiver) = bounded(prompts.len());
+
+        let num_prompts = prompts.len();
+
+        for (i, prompt) in prompts.into_iter().enumerate() {
+            let sender = sender.clone();
+            let client = self.client.clone();
             let req = self.generate_embedding_request(&prompt.to_string())?;
-            let res = self.client.execute(req).await?;
-            let res = res.json::<OpenAIEmbeddingResponse>().await?;
-            embeddings.push(res);
+
+            tokio::spawn(async move {
+                let res = client.execute(req).await.map_err(|e| e.to_string())?;
+                let response = res.json::<OpenAIEmbeddingResponse>().await.map_err(|e| e.to_string())?;
+                sender.send((i, response)).unwrap();
```

Comment:
This could look something like `sender.send((i, response)).await;` using tokio channels

Reply:
Edit: Actually, the current method of error handling within the spawned tasks can be problematic. It might be better to send back a Result type over the channel, so the receiver can decide how to handle errors.
```rust
tokio::spawn(async move {
    let result = async {
        let res = client.execute(req).await.map_err(|e| EmbeddingError::RequestError(e.to_string()))?;
        let response = res.json::<OpenAIEmbeddingResponse>().await.map_err(|e| EmbeddingError::ResponseError(e.to_string()))?;
        Ok(response)
    }.await;

    // Send back the result (success or error) via the channel.
    sender.send((i, result)).await.expect("Failed to send over channel");
});
```

Reply:
I just added simple ```format!``` errors but I was wondering did you want to create some ```thiserror``` for the EmbeddingError like in your example?

Reply:
`format!` should be fine for now! I just made up some generic type there for example purposes. Tbh Rust error handling still confuses me, so I just try to use `anyhow` for everything. In the future my plan is to clean up Orca's error handling.